File: LeetCode_Solution/LeetCode_125/Solution.js
// Problem Link: https://leetcode.com/problems/valid-palindrome/

/**
 * Check if a string is a palindrome or not
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 * 
 * @param {string} s
 * @return {boolean}
 */

function isPalindrome(s) {
  s = s.replace(/[^a-zA-Z0-9]/g, '').toLowerCase();
  let left = 0;
  let right = s.length - 1;
  while (left < right) {
    if (s[left] !== s[right]) {
      return false;
    }
    left++;
    right--;
  }
  return true;
}

console.log(isPalindrome("Hello "));

File: LeetCode_Solution/LeetCode_125/Solution.java
// Problem Link: https://leetcode.com/problems/valid-palindrome/

package LeetCode_125;

public class Solution {
  /*
   * Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.
   * 
   * Time Complexity: O(n)
   * Space Complexity: O(1)
   */

  public static boolean isPalindrome(String str, int start, int end) {
    if (start >= end)
      return true;
    if (str.charAt(start) != str.charAt(end))
      return false;
    return isPalindrome(str, start + 1, end - 1);
  }

  public static void main(String[] args) {
    String s = "sas!";
    s = s.replaceAll("[^a-zA-z0-9]", "");
    s = s.toLowerCase();
    System.out.println(isPalindrome(s, 0, s.length() - 1));
  }
}
